chore: add postgres template caching for tests

hugodutka opened the pull request:
This PR is the first in a series aimed at closing [#15109](https://github.com/coder/coder/issues/15109).

### Changes

- **Template Database Creation:**  
  `dbtestutil.Open` now has the ability to create a template database if none is provided via `DB_FROM`. The template database’s name is derived from a hash of the migration files, ensuring that it can be reused across tests and is automatically updated whenever migrations change.

- **Optimized Database Handling:**  
  Previously, `dbtestutil.Open` would spin up a new container for each test when `DB_FROM` was unset. Now, it first checks for an active PostgreSQL instance on `localhost:5432`. If none is found, it creates a single container that remains available for subsequent tests, eliminating repeated container startups.

These changes address the long individual test times (10+ seconds) reported by some users, likely due to the time Docker took to start and complete migrations.

hugodutka committed:
utilities for creating postgres databases from templates

hugodutka committed:
tests

hugodutka committed:
update comments

hugodutka committed:
update comments

hugodutka committed:
remove DB_FROM from test-postgres

hugodutka committed:
only initialize dbmem if needed

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
-
-		dbName = "ci" + dbName
-		_, err = db.Exec("CREATE DATABASE " + dbName + " WITH TEMPLATE " + os.Getenv("DB_FROM"))
+		defer cleanupConn.Close()
+		_, err = cleanupConn.Exec("DROP DATABASE " + dbName + ";")
```
Won't these panic if `cleanupConn` is nil? The `err` check above does not exit the function

hugodutka replied:
good catch!

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+		cleanupConn, err := sql.Open("postgres", cleanupDbURL)
 		if err != nil {
-			return "", nil, xerrors.Errorf("generate db name: %w", err)
+			_, _ = fmt.Fprintf(os.Stderr, "cleanup database %q: failed to connect to postgres: %s\n", dbName, err.Error())
```
Sad we do not pass in `testing.T` to properly log this.

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+// Open creates a new PostgreSQL database instance.
+// If there's a database running at localhost:5432, it will use that.
+// Otherwise, it will start a new postgres container.
+func Open(opts ...OpenOption) (string, func(), error) {
```
Should this be `Open(t *testing.T, ...)` and not return a `cleanup` func?

Then we do `t.Cleanup(close())` inside this? I know it is more refactoring :cry:

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+			panic(err)
+		}
+	}()
+	_, err = tx.Exec("SELECT pg_advisory_xact_lock(2137)")
```
is 2137 just an arbitrary number? If so, can we drop a comment on it?

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+// If templateDBName is empty, it will create a new template database based on
+// the current migrations, and name it "tpl_<migrations_hash>". Or if it's
+// already been created, it will use that.
+func createDatabaseFromTemplate(connParams ConnectionParams, newDBName string, templateDBName string) error {
```
Nice work thinking of the edge cases!

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+		return container{}, nil, xerrors.Errorf("create pool: %w", err)
 	}
 
 	tempDir, err := os.MkdirTemp(os.TempDir(), "postgres")
 	if err != nil {
```
tempDir is not used if the resource already exists

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+			Port:     port,
+			Resource: resource,
+			Pool:     pool,
+		}, func() {
+			_ = pool.Purge(resource)
+			_ = os.RemoveAll(tempDir)
+		}, nil
```
If another process is reusing the same container, does this purge kill that second container?

Is the resource existence check for when you run `go test` in 2 different terminals?

hugodutka replied:
If another process is reusing the container there'd be only one container, and the purge would kill it.

> Is the resource existence check for when you run go test in 2 different terminals?

Yes.

Emyrk replied:
If test 1 calls purge while test 2 is running. That breaks test 2 right?

hugodutka replied:
That's correct. It would have to be a named container though, and the name would have to be the same in both runs. The test could work around this by generating a unique name per test run.

hugodutka replied:
Note: no tests currently use named containers because I introduced them in this PR :)

Emyrk commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+	return container{
+			Host:     host,
+			Port:     port,
+			Resource: resource,
+			Pool:     pool,
+		}, func() {
+			_ = pool.Purge(resource)
+			_ = os.RemoveAll(tempDir)
+		}, nil
+}
+
+// OpenContainerized creates a new PostgreSQL server using a Docker container.  If port is nonzero, forward host traffic
+// to that port to the database.  If port is zero, allocate a free port from the OS.
+func OpenContainerized(opts DBContainerOptions) (string, func(), error) {
+	container, containerCleanup, err := openContainer(opts)
+	defer func() {
+		if err != nil {
+			containerCleanup()
+		}
+	}()
+	if err != nil {
+		return "", nil, xerrors.Errorf("open container: %w", err)
+	}
+	dbURL := ConnectionParams{
+		Username: "postgres",
+		Password: "postgres",
+		Host:     container.Host,
+		Port:     container.Port,
+		DBName:   "postgres",
+	}.DSN()
 
 	// Docker should hard-kill the container after 120 seconds.
-	err = resource.Expire(120)
+	err = container.Resource.Expire(120)
 	if err != nil {
 		return "", nil, xerrors.Errorf("expire resource: %w", err)
 	}
 
-	pool.MaxWait = 120 * time.Second
+	container.Pool.MaxWait = 120 * time.Second
```
Might be useful on the other docker create. If a test is `ctl+c` the cleanup will not run.

Although I do recall you mentioning the db **should** stay up after tests are done.

Emyrk reviewed (commented):
Overall I like the direction for this.

The container staying up though I am more hesitant of. It just feels strange to leave open a container for running tests. But I also see the value in keeping it open :man_shrugging:.

I won't block on that. I am ok with keeping it, and letting people complain if it becomes an issue.

hugodutka commented on coderd/database/dbtestutil/postgres.go:
```diff
@@ -16,119 +25,452 @@ import (
+		}
+	}
+	if err != nil && shouldOpenContainer {
+		// If there's no database running on the default port, we'll start a
+		// postgres container. We won't be cleaning it up so it can be reused
+		// by subsequent tests. It'll keep on running until the user terminates
+		// it manually.
```
The container actually uses ramdisk so there's no volume. Cleaning it up would be also hard when you're running more than 1 test. You'd like to share it among the tests in that run, but clean up only once they all finish. And if you're running 2 `go test` processes at once, you'd need to make sure one doesn't kill the container while the other is running.

hugodutka committed:
address review comments

hugodutka committed:
pass testing.T to dbtestutil.Open

hugodutka commented:
@Emyrk I addressed your comments. I don't think leaving a testing container open is an issue - that's how `make test-postgres` already works. Let's see how people react.

hugodutka committed:
format

hugodutka committed:
convert one more loop to use retry

Emyrk reviewed (approved):
Approving. If people end up objecting about the docker container remaining, we can tweak.

Approving assuming CI does not break either :smile:
